internal/garbagecollection/garbagecollector.go

```diff
@@ -0,0 +1,94 @@
+package garbagecollection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/metadata"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/operator-framework/rukpak/api/v1alpha2"
+)
+
+var _ manager.Runnable = (*GarbageCollector)(nil)
+
+// GarbageCollector is an implementation of the manager.Runnable
+// interface for running garbage collection on the Catalog content
+// cache that is served by the catalogd HTTP server. It runs in a loop
+// and will ensure that no cache entries exist for Catalog resources
+// that no longer exist. This should only clean up cache entries that
+// were missed by the handling of a DELETE event on a Catalog resource.
+type GarbageCollector struct {
+	CachePath      string
+	Logger         logr.Logger
+	MetadataClient metadata.Interface
+	Interval       time.Duration
+}
+
+// Start will start the garbage collector. It will always run once on startup
+// and loop until context is canceled after an initial garbage collection run.
+// Garbage collection will run again every X amount of time, where X is the
+// supplied garbage collection interval.
+func (gc *GarbageCollector) Start(ctx context.Context) error {
```

Comment:
What's the use of garbage collector running at specific internal of time? Can this be a part of finalizer, where all cached bundles belonging to a particular BD are deleted at once. 

Running a GC could turn out to be more expensive than caching a couple more bundles belonging to the same BD with different image digests. 

Referring back to my previous comment on having unpack and bundle cache - we can also consolidate them into one, if we don't do the GC and are able to pass on the right unpack location path to the handler to do the next steps in processing and creating helm chart.

Reply:
It's been long enough since I wrote this in catalogd that I can't remember the exact reason why we needed to implement a garbage collector, so it being present in this PR is a result of trying to achieve functional parity with catalogd.

I do believe catalogd uses a finalizer to also ensure cleanup is performed but I missed adding that functionality in this PR.

I don't disagree that this might be unnecessary, but in the interest of ensuring we are making these optimizations everywhere they need to be I think we should:
- Leave the garbage collector in place for this PR
- After this PR merges, create an epic/issues to update this package to work for both catalogd and operator-controller use cases
- Make any necessary optimizations to the implementation (one location for changes, benefits multiple projects)

Reply:
Additionally, the garbage collector doesn't _need_ to be copied over into operator-controller if it isn't needed. The real intention for this PR is to get the direct image registry client functionality in place and able to be used in operator-controller via the `source` pkg. 

If all we care about is getting that code in I can strip all the other changes from this PR and only add the functionality for unpacking by talking directly to an image registry and we handle everything else in operator-controller as we see fit.

Reply:
I'd suggest removing the garbage collector, and keeping the PR confined to adding direct image registry client with a finalizer to clean up bundle contents - which is the bare minimum requirement to unblock currently. GC or any other improvements to maintain common code b/w catalogd and o-c/rukpak can be handled in follow ups based on future discussions.

Reply:
Garbage collection removed and finalizer logic added in https://github.com/operator-framework/rukpak/pull/874/commits/59925ff9d2361a49b8a824da8b644e9f0365ce61